Add clang-17 to CI

bernhardmgruber opened the pull request:
Fixes: #2169

bernhardmgruber committed:
Add clang-17 to CI

Fixes: #2169

j-stephan commented:
Okay, this is definitely the last feature PR going into 1.0. If some other compiler releases today or tomorrow it will not be included, nor will any fixes for this compiler.

I'll submit a PR which cherry-picks the commits since the creation of the 1.0 branch; afterwards we should be ready for the release.

bernhardmgruber commented:
> Okay, this is definitely the last feature PR going into 1.0. If some other compiler releases today or tomorrow it will not be included, nor will any fixes for this compiler.

I understand :)

SimeonEhrig commented:
Clang 17 is not official supported by the `alpaka-job-coverage-library`. Please open an issue: https://github.com/alpaka-group/alpaka-job-matrix-library/issues

bernhardmgruber commented:
> Clang 17 is not official supported by the `alpaka-job-coverage-library`. Please open an issue: https://github.com/alpaka-group/alpaka-job-matrix-library/issues

Wait. Now adding a new compiler version also requires to add issues, PRs and commits to other repositories? That seems a bit .. complex.

Anyhow, here is the issue: https://github.com/alpaka-group/alpaka-job-matrix-library/issues/39

SimeonEhrig commented:
*

> > Clang 17 is not official supported by the `alpaka-job-coverage-library`. Please open an issue: https://github.com/alpaka-group/alpaka-job-matrix-library/issues
> 
> Wait. Now adding a new compiler version also requires to add issues, PRs and commits to other repositories? That seems a bit .. complex.
> 
> Anyhow, here is the issue: [alpaka-group/alpaka-job-matrix-library#39](https://github.com/alpaka-group/alpaka-job-matrix-library/issues/39)

Yes, but only because we share the CI with cupla, vikunja and PIConGPU.

bernhardmgruber commented:
I am moving this off to after the 1.0 release, since the remaining failures are with `-Werror` and CUDA mode. alpaka compiles fine with clang-17 outside those use cases.

bernhardmgruber commented:
There are failures with generating ptx again:
```
ptxas /tmp/randomStrategies-sm_61-768a17.s, line 14415; fatal   : Parsing error near '.': syntax error
ptxas fatal   : Ptx assembly aborted due to errors
clang++: error: ptxas command failed with exit code 255 (use -v to see invocation)
```
I think we had this before somewhere?

j-stephan commented:
Yes, it is this one: https://github.com/alpaka-group/alpaka/issues/1906 Glad to see this issue persisting on LLVM's side...

bernhardmgruber committed:
Rename lambda captures to workaround warnings

clang warns that the names of the captures shadow the outside variables.
E.g.:

QueueUniformCudaHipRt.hpp:215:57: error: declaration shadows a local variable [-Werror,-Wshadow-uncaptured-local]
  215 |                 auto f = queue.m_callbackThread.submit([data = std::move(data)] { data->t(); });
      |                                                         ^

bernhardmgruber committed:
Exclude clang CUDA Debug builds from the CI

They fail with:
ptxas /tmp/randomStrategies-sm_61-768a17.s, line 14415; fatal   : Parsing error near '.': syntax error
ptxas fatal   : Ptx assembly aborted due to errors
clang++: error: ptxas command failed with exit code 255 (use -v to see invocation)

bernhardmgruber commented:
> Yes, it is this one: #1906 Glad to see this issue persisting on LLVM's side...

Thx, I excluded clang-17 as CUDA compiler in Debug mode, as we do for older clang versions.

j-stephan commented:
You also need to update `README.md` :-)

bernhardmgruber committed:
Add clang-17 to README.md

CUDA/HIP/SYCL is marked as untested for now.

bernhardmgruber commented:
> You also need to update `README.md` :-)

Right. Added clang-17 for all CPU backends, because I did not have the chance to test what combinations of nvcc+clang and clang CUDA work.

SimeonEhrig commented on README.md:
```diff
@@ -65,17 +65,17 @@ Supported Compilers
+| std::thread                                                                    | :white_check_mark:                              | :white_check_mark:                              | :white_check_mark:                          | :white_check_mark:     | :white_check_mark:                                         | :white_check_mark:                                    | :white_check_mark:           ...
+| TBB                                                                            | :white_check_mark:                              | :white_check_mark:                              | :white_check_mark:                          | :white_check_mark:     | :white_check_mark:                                         | :white_check_mark:                                    | :white_check_mark:           ...
+| CUDA (nvcc)                                                                    | :white_check_mark: <br/> (CUDA 11.0 - 12.2)[^2] | :white_check_mark: <br/> (CUDA 11.4 - 12.0)[^2] | :white_check_mark: <br/> (CUDA 12.0 - 12.2) | :x:                    | :white_check_mark: <br/> (CUDA 11.0-11.2; 11.6 - 12.0)[^2] | :white_check_mark: <br/> (CUDA 11.2, 11.6 - 12.0)[^2] | :white_check_mark: <br/> (CU ...
+| CUDA (clang)                                                                   | -                                               | -                                               | -                                           | :x:                    | :x:                                                        | :x:                                                   | :x:                          ...
```
Why is CUDA (clang) not supported? Clang 17 supports up to CUDA 12.1: https://llvm.org/docs/CompileCudaWithLLVM.html
And we have a CI job, which test it.

bernhardmgruber replied:
I added the range CUDA 11.0 - 12.1 then.

j-stephan replied:
According to https://github.com/llvm/llvm-project/blob/llvmorg-17.0.0/clang/include/clang/Basic/Cuda.h 11.8 is the last "fully supported" version, 12.1 is "partially supported". So far we always used the "fully supported" version in our matrix.

bernhardmgruber replied:
So CUDA 11.0 - 11.8 then?

j-stephan replied:
If we want to stay consistent, yes.

bernhardmgruber replied:
Alright.
